quinn-proto/src/connection.rs

```diff
@@ -151,7 +153,12 @@ where
 
     streams: Streams,
     /// Surplus remote CIDs for future use on new paths
-    rem_cids: Vec<IssuedCid>,
+    ///
+    /// Sequence numbers start at `self.first_unretired_cid` and are contiguous, including `None`
+    /// slots. Always exactly REM_CID_COUNT elements long. Maintaining a fixed-size window allows us
+    /// to easily discard duplicate and obsolete NEW_CONNECTION_ID frames fast and with constant
+    /// memory.
+    rem_cids: VecDeque<Option<(ConnectionId, ResetToken)>>,
```

Comment:
I find the design here quite complex. Did you think about storing the sequence numbers in the deque along with the other state instead (it would become `VecDeque<(u64, Option<(ConnectionId, ResetToken)>)>`, I guess)? This should obviate the need for the separate `first_unretired_cid`. It would take a bit of extra memory, but 64 every bytes in the heap probably won't kill us.

(Perhaps `(u64, Option<(ConnectionId, ResetToken)>)` should become a proper struct instead, in which case we can also derive `Default` for it.)

Reply:
We're specifically trying to represent a continuous sliding window. I'm hence not a fan of explicit sequence numbers, because they allow a far larger number of invalid states to be represented. The latest iteration of this is hopefully a bit clearer.